language/bytecode-verifier/src/friends.rs

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module contains verification of usage of dependencies for modules
+use diem_types::vm_status::StatusCode;
+use vm::{
+    access::ModuleAccess,
+    errors::{Location, PartialVMError, PartialVMResult, VMResult},
+    file_format::CompiledModule,
+};
+
+pub fn verify_module(module: &CompiledModule) -> VMResult<()> {
+    verify_module_impl(module).map_err(|e| e.finish(Location::Module(module.self_id())))
+}
+
+fn verify_module_impl(module: &CompiledModule) -> PartialVMResult<()> {
+    // cannot make friends with the module itself
+    let self_index = module.self_handle_idx();
+    if module.immediate_friends().contains(&self_index) {
+        return Err(PartialVMError::new(
+            StatusCode::INVALID_FRIEND_DECL_WITH_SELF,
+        ));
+    }
+
+    // cannot make friends with modules outside of the account address
```

Comment:
I still really don't agree with this check conceptually. But I'm okay I suppose with it being here now, as we can remove it later... 
@sblackshear any thoughts?

Reply:
I agree on both points. Perhaps an additional comment explaining that this is a policy decision rather than a technical one + that we may consider lifting this limitation in the future.